lib/gcloud/search/api_client.rb

```diff
@@ -0,0 +1,128 @@
+    # discoverable, initialization of this class in Connection should be
+    # replaced with the Google API Client.
+    class APIClient #:nodoc:
+      attr_accessor :authorization, :connection
+
+      ##
+      # Creates a new APIClient instance.
+      def initialize _options
+        @connection = Faraday.default_connection
+      end
+
+      def discovered_api name, version
+        DiscoveredApi.new name, version
+      end
+
+      def execute options
+        api_method = options[:api_method]
+        uri = api_method[:uri]
+        params = options[:parameters]
+        uri.gsub! "{projectId}", params.delete(:projectId) if params[:projectId]
+        uri.gsub! "{indexId}", params.delete(:indexId) if params[:indexId]
+        uri.gsub! "{docId}", params.delete(:docId) if params[:docId]
+        run api_method[:method], uri, options
+      end
+
+      def inspect
+        "#{self.class}(#{@project})"
+      end
+
+      protected
+
+      ##
+      # Return type for APIClient#discovered_api
+      class DiscoveredApi
+        def initialize name, version
+          @name = name
+          @version = version
+        end
+
+        def indexes
+          IndexResourcePath.new @name, @version, "indexes", "indexId"
+        end
+
+        def documents
+          ResourcePath.new @name,
+                           @version,
+                           "indexes/{indexId}/documents",
+                           "docId"
+        end
+      end
+
+      ##
+      # Return type for DiscoveredApi http verb methods
+      class ResourcePath
+        def initialize api_name, api_version, resource_root, resource_id_param
+          @root = "https://#{api_name}.googleapis.com/#{api_version}" \
+                  "/projects/{projectId}/#{resource_root}"
+          @resource_id_param = resource_id_param
+        end
+
+        def create
+          api_method :post
+        end
+
+        def delete
+          api_method :delete, "/#{@resource_id_param}"
+        end
+
+        def get
+          api_method :get, "/#{@resource_id_param}"
+        end
+
+        def list
+          api_method :get
+        end
+
+        def api_method method, path = nil
+          { method: method, uri: "#{@root}#{path}" }
+        end
+      end
+
+      ##
+      # Special-case return type for DiscoveredApi http search verb method
+      class IndexResourcePath < ResourcePath
+        def search
+          api_method :get, "/search"
+        end
+      end
+
+      def run method, uri, options = {}
+        if authorization.nil?
+          @connection.send method do |req|
+            req.url uri
+            req.params = options[:parameters] if options[:parameters]
+            req.body = options[:body_object] if options[:body_object]
+          end
+        else
+          # TODO: Support POST, DELETE
+          options[:uri] = uri
+          options[:connection] = @connection
+          authorization.fetch_protected_resource options
```

Comment:
@blowmage Have you already figured out which additional options (or other method calls) are required with Signet's `fetch_protected_resource` to support POST, DELETE in production?